File: UIPreviewInteraction/UIPreviewInteractionObjCSample/UIPreviewInteractionObjCSample/PreviewInteractionMainViewController.m
//
//  PreviewInteractionMainViewController.m
//  UIPreviewInteractionObjCSample
//
//  Created by nikolay.andonov on 10/17/16.
//  Copyright © 2016 nikolay.andonov. All rights reserved.
//

#import "PreviewInteractionMainViewController.h"
#import "PreviewInteractionPopViewController.h"

@interface PreviewInteractionMainViewController () <UIPreviewInteractionDelegate>
@property (weak, nonatomic) PreviewInteractionPopViewController *popViewController;
@property (strong, nonatomic) UIPreviewInteraction *previewInteraction;

@end

@implementation PreviewInteractionMainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.previewInteraction = [[UIPreviewInteraction alloc] initWithView: self.view];
    self.previewInteraction.delegate = self;

}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([segue.destinationViewController isKindOfClass:[PreviewInteractionPopViewController class]]) {
        self.popViewController = segue.destinationViewController;
    }
}

#pragma mark - UIPreviewInteractionDelegate

- (void)previewInteraction:(UIPreviewInteraction *)previewInteraction didUpdatePreviewTransition:(CGFloat)transitionProgress ended:(BOOL)ended {
    
    if (!self.popViewController) {
        [self performSegueWithIdentifier:@"InteractionSegue" sender:nil];
    }
    self.popViewController.progressLabel.text = [NSString stringWithFormat:@"%.0f%%", transitionProgress * 100];
    
    if (ended) {
        self.popViewController.statusLabel.text = @"Peek!";
    }
}

- (void)previewInteraction:(UIPreviewInteraction *)previewInteraction didUpdateCommitTransition:(CGFloat)transitionProgress ended:(BOOL)ended {
    
    self.popViewController.animator.fractionComplete = transitionProgress;
    self.popViewController.progressLabel.text = [NSString stringWithFormat:@"%.0f%%", transitionProgress * 100];
    if (ended) {
        self.popViewController.statusLabel.text = @"Pop!";
    }
}

- (void)previewInteractionDidCancel:(UIPreviewInteraction *)previewInteraction {
    
    self.popViewController.statusLabel.text = @"";
    [self.popViewController dismissViewControllerAnimated:YES completion:nil];
}

@end

File: UIPreviewInteraction/UIPreviewInteractionSwiftSample/UIPreviewInteractionSwiftSample/PreviewInteractionMainViewController.swift
//
//  PreviewInteractionMainViewController.swift
//  UIPreviewInteractionSwiftSample
//
//  Created by nikolay.andonov on 10/17/16.
//  Copyright © 2016 nikolay.andonov. All rights reserved.
//

import UIKit

class PreviewInteractionMainViewController: UIViewController, UIPreviewInteractionDelegate {

    private var previewInteraction: UIPreviewInteraction!
    weak private var popViewController: PreviewInteractionPopViewController?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        previewInteraction = UIPreviewInteraction(view: view)
        previewInteraction.delegate = self
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if let viewController = segue.destination as? PreviewInteractionPopViewController {
            popViewController = viewController
        }
    }

    // MARK: - UIPreviewInteractionDelegate

    func previewInteraction(_ previewInteraction: UIPreviewInteraction, didUpdatePreviewTransition transitionProgress: CGFloat, ended: Bool) {
        
        if popViewController == nil {
            performSegue(withIdentifier: "InteractionSegue", sender: nil)
        }
        guard let popViewController = popViewController else {fatalError()}
        popViewController.progressLabel.text = String(format: "%.0f%", transitionProgress * 100)
        
        if ended {
            popViewController.statusLabel.text = "Peek!"
        }
    }
    
    func previewInteraction(_ previewInteraction: UIPreviewInteraction, didUpdateCommitTransition transitionProgress: CGFloat, ended: Bool) {
        
        guard let popViewController = popViewController else {fatalError()}
        popViewController.animator.fractionComplete = transitionProgress
        popViewController.progressLabel.text = String(format: "%.0f%", transitionProgress * 100)
        if ended {
            popViewController.statusLabel.text = "Pop!"
        }
    }
    
    func previewInteractionDidCancel(_ previewInteraction: UIPreviewInteraction) {
        dismiss(animated: true) {
            self.popViewController?.statusLabel.text = ""
        }
    }
}
